pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ProducerImpl.java

```diff
@@ -484,7 +497,105 @@ void ackReceived(ClientCnx cnx, long sequenceId, long ledgerId, long entryId) {
         }
     }
 
-    private static final class OpSendMsg {
+    /**
+     * Checks message checksum to retry if message was corrupted while sending to broker. Recomputes checksum of the
+     * message header-payload again.
+     * <ul>
+     * <li><b>if doesn't match with existing checksum</b>: it means message was corrupt while sending to broker. So, set
+     * new checksum and resend message</li>
+     * <li><b>if matches with existing checksum</b>: it means message is already corrupt and can't retry again. So, fail
+     * send-message by failing callback</li>
+     * </ul>
+     * 
+     * @param cnx
+     * @param sequenceId
+     */
+    protected void recoverChecksumError(ClientCnx cnx, long sequenceId) {
+        OpSendMsg op = null;
+        synchronized (this) {
+            op = pendingMessages.peek();
+            if (op == null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] [{}] Got send failure for timed out msg {}", topic, producerName, sequenceId);
+                }
+            } else {
+                long expectedSequenceId = op.sequenceId;
+                if (sequenceId == expectedSequenceId) {
+                    boolean isUpdated = updateChecksumIfRequire(op);
+                    if (!isUpdated) {
+                        op.callback.sendComplete(
+                                new PulsarClientException.ChecksumException("Checksum failded on corrupt message"));
+                        return;
+                    }
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("[{}] [{}] Corrupt message is already timed out {}", topic, producerName, sequenceId);
+                    }
+                }
+            }
+        }
+        // close connection and let producer resend pending-messages
+        cnx.ctx().close();
+    }
+    
+    /**
+     * Computes checksum again and verifies it against existing checksum. 
+     * If checksum doesn't match then updates newly computed checksum {@link OpSendMsg}. 
+     * 
+     * @param op
+     * @return isUpdated: returns true only if checksum is updated in {@link OpSendMsg}
+     */
+    protected boolean updateChecksumIfRequire(OpSendMsg op) {
+        
+        if (op.cmd instanceof DoubleByteBuf) {
```

Comment:
Here, to avoid creation of new ByteBuf we modify same DoubleByteBuf of the message with newly computed checksum.
However, while message creation if we see memory-leak then we create ```SimpleLeakAwareByteBuf``` or ```AdvancedLeakAwareByteBuf``` (based on ```ResourceLeak Level```) instead   ```DoubleByteBuf```. So, should we keep this check.